### PR TITLE
"Offline" Capable

### PR DESCRIPTION
**Pull request by rgaiacs**
Some discussion at [shell-extra](https://github.com/swcarpentry/shell-extras/) mention the desired for the lessons be "offline" capable.

# Description

Links in the lesson are of the form `http://swcarpentry.github.io/LESSON-NAME/XX-EPISODE-NAME/`. The links works when the lesson is being server by a web server because the server know that it should return the file `_site/XX-EPISODE-NAME/index.html` **but will not work if viewing the lesson over [file URI scheme](https://en.wikipedia.org/wiki/File_URI_scheme)** because `file:///PATH/TO/LESSON/ROOT/_site/XX-EPISODE-NAME/` is not the same as `file:///PATH/TO/LESSON/ROOT/_site/XX-EPISODE-NAME/index.html`.

# Proposal

Adopt `http://swcarpentry.github.io/LESSON-NAME/XX-EPISODE-NAME/index.html` instead of `http://swcarpentry.github.io/LESSON-NAME/XX-EPISODE-NAME/`.

# Benefits

As mention before, read the lesson without running a web server.

# Drawback

Except for the HTML files be `10 times number of links` characters longer, **no change**.

# Implementation

Most of the links are compose, like in `_includes/syllabus.html`, by

~~~
<a href="{{ page.root }}{{ episode.url }}">{{ episode.title }}</a>
~~~

where `episode.url` is defined by `permalink` on

~~~
collections:
  episodes:
    output: true
    permalink: /:path/
~~~

at `_config.yml`. We can change

~~~
    permalink: /:path/
~~~

to

~~~
    permalink: /:path/index.html
~~~

and accomplished the desired change.

# TODO

- [x] List other links that need to be updated and do similar change
- [ ] Migrate the pull request to https://github.com/swcarpentry/styles/

# Related Work

@vuw-ecs-kevin created https://github.com/vuw-ecs-kevin/offline-capable-lesson/ to prototype a alternative solution. See https://github.com/vuw-ecs-kevin/offline-capable-lesson/commit/66442fcdf40f967302bc714c4b693f9355eb4061 for the implementation.

# Request to comment

@vuw-ecs-kevin @brandoncurtis @ChristinaLK

### CODE REVIEW

**Comment by vuw-ecs-kevin** [comment]
Just to say that In the way I've always thought of this, the PROPOSAL is NOT to

> Adopt `http://swcarpentry.github.io/LESSON-NAME/XX-EPISODE-NAME/index.html` instead of
> `http://swcarpentry.github.io/LESSON-NAME/XX-EPISODE-NAME/.`

but merely to ALLOW FOR the specification of a default filename in the links within the lessons.

In the way I developed the idea to allow for a correct navigation of a 
"fully offline capable" Lesson, it would be up to the person rendering the 
Lesson, to supply a value for a lesson-wide variable, for example, `{{ site.index }}`,
specified in `_config.yml` as:

```
# Filename matching webserver default, for offline browsing
index: "index.html"
```
with which to populate instance of the variable within a Lesson's source,
however if the value for that variable is empty, links within a Lesson wold
appear just as they do now, so

`http://swcarpentry.github.io/LESSON-NAME/XX-EPISODE-NAME/`

which I believe to be a good thing.

In other words the default value for the variable would be empty, and it would
be up to the person rendering it, to place `index.html`, or whatever value they
thought necessary, prior to their rendering.

Kevin M. Buckley

eScience Consultant
School of Engineering and Computer Science
Victoria University of Wellington
New Zealand

**Comment by vuw-ecs-kevin** [comment]
As per Raniere's (**rgaiacs**) comment 

https://github.com/swcarpentry/shell-extras/pull/38#issuecomment-324835530

to submit a PR to the swcarpentry:styles repository, I have just created a PR

https://github.com/swcarpentry/styles/pull/166

that  shows the deployment of my original *solution* which, I believe fixes up all of the 
links within the current Lesson content, so episodes and auxilliary content.

**Comment by rgaiacs** [comment]
@vuw-ecs-kevin Again, sorry for the delay to work on this.

I did some work, rebased my changes, squash some commits, ... and this looks good to me except for a documentation about the use of Jekyll tag `link`. This is a more robust solution because if we decide to remove this we can just change the `permalink` variable. And as I mention on the first comment, no drawback.

I'm planning to merge this for the next release.

**Comment by vuw-ecs-kevin** [comment]
> @vuw-ecs-kevin Again, sorry for the delay to work on this.

Please, don't **ever** feel the need to apologise to me for any perceived 
delay in projects such as this: and certainly, never apologise twice!

I haven't experienced any delays that have prevented me from doing 
anything further with this, or any other related project.

Given the collaborative nature of such projects, suggestions such as the one 
I made here are rarely likely to be the exact soultion required, especially when 
they are "solutions" to questions that may not even have been asked before, 
by the people with the greatest amount of knowledge/experience of where a
project is heading, and I am equally sure that everyone who comes to use the 
Carpentries, knows just how much has to go on behind the scenes, just to make 
this all happen and be of benefit.

What is also worth remmebering, when discussing "delays" in such projects, is 
that, because of the open nature of the suggestions to projects, anyone who wishes 
to take a particular suggestion on board, may do so, simply by applying the initial 
patches locally, until such time as the "official" set of patches is reached, so there
is no delay. Furthermore, even if a suggestion is not taken up, then those who still
want to use it may still do so.

Yours, looking forward to taking a look at how you did it,
Kevin

**Comment by vuw-ecs-kevin** [comment]
> @vuw-ecs-kevin Again, sorry for the delay to work on this.
> 
> I did some work, rebased my changes, squash some commits, ... and this looks good to me except for a documentation about the use of Jekyll tag link. This is a more robust solution because if we decide to remove this we can just change the permalink variable. And as I mention on the first comment, no drawback.
> 
> I'm planning to merge this for the next release.

I just took the diffs, and applied them to a clone of the SWC lesson-example repo, 
and then ran a `make site`.

When viewing the renderdd lesson as via a `file:///path/to/lesson-example/_site/index.html`, 
none of the items in the menu bar that the diffs affect, have ended up with an `index.html` 
added onto the end of the link target? 

So for example, the NavBar's "Code of Conduct" menu item's target is still:

`file:///path/to//lesson-example/_site/conduct/`

Shouldn't it now have an `index.html'` there : or don't your patches here address 
the "offline capable" issue but merely tidy up some internal code ?

Apologies if I have missed something,
Kevin

**Comment by rgaiacs** [comment]
> When viewing the renderdd lesson as via a file:///path/to/lesson-example/_site/index.html,
none of the items in the menu bar that the diffs affect, have ended up with an index.html
added onto the end of the link target?

Small issue when I was rebasing my commits. 15b61b4 has the correct link for the episodes, e.g. `./01-design/index.html` instead of `./01-design/` but it has the wrong link for pages in the root directory as you mention. Thanks for the review.

d0a62bd fix the issue. Pages at the root directory have the correct link, e.g. `./CONDUCT.html`. Pages at the root doesn't receive a directory so they don't need the `index.html`.

~~~
$ git log --oneline -2 
* d0a62bd Fix root pages
* 15b61b4 Add changes to example files
$ make site
jekyll build
Configuration file: /home/raniere/software-carpentry/swc-lesson-example/_config.yml
            Source: /home/raniere/software-carpentry/swc-lesson-example
       Destination: /home/raniere/software-carpentry/swc-lesson-example/_site
 Incremental build: disabled. Enable with --incremental
      Generating... 
                    done in 0.544 seconds.
 Auto-regeneration: disabled. Use --watch to enable.
$ cat _site/index.html
...
      <ul class="nav navbar-nav">

	
        <li><a href="./CONDUCT.html">Code of Conduct</a></li>

        
	
        <li><a href="./setup.html">Setup</a></li>

        
        <li class="dropdown">
          <a href="./" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Episodes <span class="caret"></span></a>
          <ul class="dropdown-menu">
            
            <li><a href="./01-design/index.html">Lesson Design</a></li>
            
            <li><a href="./02-tooling/index.html">GitHub, Markdown, and Jekyll</a></li>
            
            <li><a href="./03-organization/index.html">Lesson Organization</a></li>
            
            <li><a href="./04-formatting/index.html">Formatting</a></li>
            
            <li><a href="./05-checking/index.html">Checking and Previewing</a></li>
            
            <li><a href="./06-rmarkdown-example/index.html">Using RMarkdown</a></li>
            
            <li><a href="./07-maintenance/index.html">Maintenance</a></li>
            
            <li><a href="./08-coffee/index.html">Coffee Break</a></li>
            
	    <li role="separator" class="divider"></li>
            <li><a href="./aio.html">All in one page (Beta)</a></li>
          </ul>
        </li>
	

	
	
        <li class="dropdown">
          <a href="./" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Extras <span class="caret"></span></a>
          <ul class="dropdown-menu">
            <li><a href="./reference.html">Reference</a></li>
            
            <li><a href="./about/index.html">About</a></li>
            
            <li><a href="./discuss/index.html">Discussion</a></li>
            
            <li><a href="./figures/index.html">Figures</a></li>
            
            <li><a href="./guide/index.html">Instructor Notes</a></li>
            
          </ul>
        </li>
	

	
        <li><a href="./LICENSE.html">License</a></li>
	
	<li><a href="/edit/gh-pages/index.md">Improve this page <span class="glyphicon glyphicon-pencil" aria-hidden="true"></span></a></li>
	
      </ul>
...
~~~

**Comment by anelda** [comment]
Hi all, thanks very very much for this! It's going to be incredibly useful in most of the workshops we run in Africa! Appreciate your hard work around offline capability of lessons!
